### ISSUE TITLE
Proxy support does not work properly with CONNECT on https requests

### ISSUE THREAD

**Issue by gm-ghanover** [axios/axios#4531]
#### Describe the bug
Using Axios 0.26.1 behind a corporate HTTP tunneling proxy to an HTTPS destination, once the proxy connection is made, a POST method is issued, rather than a CONNECT to establish the TLS tunnel to the destination host where the POST should be sent. Result is **Error: socket hang up**, as the proxy expects a CONNECT to be issued and drops the connection

Specifically, using wireshark to monitor the activity of the script goes as follow:
> TCP SYN
< TCP SYN, ACK
> TCP ACK
> HTTP POST
< TCP FIN, ACK
> TCP ACK
![axios_pcap](https://user-images.githubusercontent.com/15218042/158659772-3d084e3b-8ab1-4f99-8c4d-0a54ed697ed4.png)

If you use curl with the same environment, the packets are as follows:
> TCP SYN
< TCP SYN, ACK
> TCP ACK
> HTTP CONNECT
< HTTP Connection established
> TCP ACK
> TLS Client Hello
< TLS Server Hello
... encrypted POST and response follows
![curl_pcap](https://user-images.githubusercontent.com/15218042/158659774-5c84d76a-3370-4acb-9a5d-f9270de12bf7.png)

#### To Reproduce

```bash
export http_proxy="http://user:password@10.123.173.125:80"
export https_proxy="http://user:password@10.123.173.125:80"
```

```js
http.post('https://api.github.com/user/repos', {});
```

#### Expected behavior
When connecting to an HTTPS server through a tunneling HTTP proxy, the CONNECT method should be used to establish the TLS handshake before issuing the POST

#### Environment
 - Axios Version 0.26.1
 - Adapter HTTP
 - Node.js Version 14.8.0
 - OS: Windows 10 (10.0.19042.1526)

**Comment by gm-ghanover** [comment]
FWIW, https://github.com/thib3113/axios-proxy-tunnel/blob/master/src/axios-proxy-tunnel.ts is an example that works with an older version of axios, doing a modification of axiosProxyTunnel is a usable workaround

**Comment by ahszww** [comment]
This problem has existed for a long time, hope it will be solved

**Comment by srdi** [comment]
I've also faced this issue.
It seems hpagent package help to fix this issue.

Here is example of usage:

`
const axios = require('axios');
const { HttpProxyAgent, HttpsProxyAgent } = require('hpagent')

const instance = axios.create({ httpAgent, httpsAgent });

instance.post('http://example.com')
`

**Comment by Nevon** [comment]
Why is the "possible bug" label removed? Clearly the automatic proxy configuration is incorrect. If the protocol is HTTPS and there is an HTTP proxy set, the client should use the tunnel protocol by sending a CONNECT request, wait to get a 200 OK back and then send the original request on the same connection.

**Comment by Yueren-Wang** [comment]
Is there an update for this issue? which axios version should we use to workaround?

**Comment by srdi** [comment]
> Is there an update for this issue? which axios version should we use to workaround?

This worked for me:



> I've also faced this issue. It seems hpagent package help to fix this issue.
> 
> Here is example of usage:
> 
> ```
> const axios = require('axios');
> const { HttpProxyAgent, HttpsProxyAgent } = require('hpagent')
> 
> const httpAgent = new HttpProxyAgent({
>     proxy: 'proxy url'
> })
> const httpsAgent = new HttpsProxyAgent({
>     proxy: 'proxy url'
> })
> 
> const instance = axios.create({ httpAgent, httpsAgent });
> 
> instance.post('http://example.com')
> ```

**Comment by ppati000** [comment]
If anyone is interested in trying out HTTPS-over-HTTP support in Axios, feel free to check https://github.com/axios/axios/pull/5037. The implementation has some limitations (no redirects) but maybe it still fits your use cases. Would love to hear back from you!

**Comment by topquarck** [comment]
> I've also faced this issue. It seems hpagent package help to fix this issue.
> 
> Here is example of usage:
> 
> ```
> const axios = require('axios');
> const { HttpProxyAgent, HttpsProxyAgent } = require('hpagent')
> 
> const httpAgent = new HttpProxyAgent({
>     proxy: 'proxy url'
> })
> const httpsAgent = new HttpsProxyAgent({
>     proxy: 'proxy url'
> })
> 
> const instance = axios.create({ httpAgent, httpsAgent });
> 
> instance.post('http://example.com')
> ```

that didn't work with me

**Comment by sebaplaza** [comment]
Can we just use [proxy-agent](https://github.com/TooTallNate/proxy-agents/tree/main/packages/proxy-agent) to fix this ?

With this module we could have all proxy features enabled automatically.

I was having the same problem, and now everything works when i replace the agent.

> proxy: false is needed, otherwise axios is trying to overwrite the agent.

```js
const axios = require("axios").default;
const { ProxyAgent } = require("proxy-agent");

async function call() {
  const url = "https://api64.ipify.org?format=json";
  const agent = new ProxyAgent();
  const res = await axios.get(url, { httpAgent:agent, httpsAgent:agent, proxy: false });
  console.log(res.data);
}
call();
```

```sh
HTTPS_PROXY=http://proxy:3128 node example.js
```

**Comment by Nevon** [comment]
Yes, you can pass in a working agent implementation yourself. The problem is that Axios currently has a built-in implementation that is broken, so if you have the HTTPS_PROXY environment variable set it will fail by default.

**Comment by sebaplaza** [comment]
Yes @Nevon, but i think axios could use proxy-agent internally. There is no need to have a built-in implementation of this.

Even if the proxy is not very difficult to understand, there is a lot of little details that can lead to multiple kind of bugs.

**Comment by cnlab-software-ag** [comment]
Is there any progress on this issue? 

Axios not properly supporting https requests over a proxy prevents many depending products from working properly.

In my case, I can not upgrade to a recent release of n8n because they dropped the alternative library on the step to 1.x.

Please fix this as this plays a central role in corporate environments.

**Comment by schuerg** [comment]
I faced the same problem and solved it using https://github.com/gajus/global-agent

It works with with all http client libraries that internally use [http.request](https://nodejs.org/api/http.html#http_http_request_options_callback), including axios.

**Comment by jasonsaayman** [comment]
Thanks for your bug report, we will be closing this issue and ask that you please test and try re-create the issue with the latest version of Axios. If the bug is still present please open a issue. If possible and you are willing open a pull request with a proposed fix. Please see the motivation for doing this:

https://github.com/axios/axios/issues/6983
